utils/wasm-builder/src/lib.rs

```diff
@@ -208,15 +209,19 @@ fn get_rustup_nightly(selected: Option<String>) -> Option<CargoCommand> {
 			let output = Command::new("rustup").args(&["toolchain", "list"]).output().ok()?.stdout;
 			let lines = output.as_slice().lines();
 
-			let mut latest_nightly = None;
+			let mut found_version = None;
 			for line in lines.filter_map(|l| l.ok()) {
-				if line.starts_with("nightly-") && line.ends_with(&host) {
-					// Rustup prints them sorted
-					latest_nightly = Some(line.clone());
+				let version = line.trim_end_matches(&host);
+
+				if CargoCommand::new_with_args("rustup", &["run", &version, "cargo"])
+					.supports_substrate_wasm_env()
+				{
+					found_version = Some(version.into());
+					break
 				}
```

Comment:
Previously this just picked the newest nightly and now it'll pick the first toolchain which rustup returns, which (based on the order in which rustup returns the toolchains) it means that it will prioritize:

1) the non-versioned "stable" toolchain or "beta" toolchain, if compatible (and this won't necessarily be the newest stable that the user has installed),
2) the oldest compatible installed nightly,
3) the oldest compatible installed stable.

Besides the fact that this changes the previous behavior (e.g. it'll break the build on my machine as I have a lot of old nightlies installed :P) considering that we aggressively use new Rust features potentially picking the oldest installed toolchain is probably not what we want?

Maybe a better approach would be to do something like this:

1) parse the date from every installed cargo/rustc's version string (note: stable can be installed either as an explicitly versioned toolchain or just as "stable", so parsing rustup's output is not enough here),
2) pick the newest stable and the newest nightly,
3) if the stable is new enough - use that, otherwise use nightly

This would also make it not depend on the order in which `rustup` returns the toolchains (which I'm not sure if it's officially "stable").

Reply:
Yeah I know that this changes the behavior. However, this function should be called as last resort, aka when the `CARGO` env etc all are not being suitable. But yeah, I can improve the check!